### PR TITLE
Policy category - Add GL Code

### PR DESCRIPTION
**Pull request by rushatgabhane**
### Fixed Issues

$ https://github.com/Expensify/App/issues/44255
PROPOSAL: N.A.

### Tests

1. Go to a collect policy
2. Go to category page
3. Select any category
4. Verify that you see GL Code
5. Verify that you can set a GL Code in the text input.
6. Verify that clicking save sets the GL Code and navigates you back
- [x] Verify that no errors appear in the JS console

### Offline tests

1. Go offline and follow the same steps above
### QA Steps

Same as tests
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

image

image

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

image

image

</details>

<details>
<summary>iOS: mWeb Safari</summary>

image

image

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/29673073/8ba5b6d9-4b70-4a0b-904d-69128c9afe2c

</details>

<details>
<summary>MacOS: Desktop</summary>

image

</details>

### CODE REVIEW

**Comment by rushatgabhane** [comment]
Note: The API command might not work because the api isn't ready yet. Please review everything else

File: src/pages/workspace/categories/CategoryGLCodePage.tsx
Comment (rushatgabhane) [inline src/pages/workspace/categories/CategoryGLCodePage.tsx:26]:
```suggestion
    /** Collection of categories attached to a policy */
```

**Review by rushatgabhane — COMMENTED** [review]
@Gonals / @Beamanator requesting a buddy check to make sure that im going in the right direction. If all looks fine, I'll raise a PR for the other pages too.

File: src/languages/en.ts
Comment (rushatgabhane) [inline src/languages/en.ts:2179]:
```suggestion
```

File: src/languages/es.ts
Comment (rushatgabhane) [inline src/languages/es.ts:2217]:
```suggestion
```

File: src/languages/en.ts
Comment (Beamanator) [inline src/languages/en.ts:2177]: I thinkkkkkkk we often want to have capital C, but could be good to confirm that with someone other than me 😅 

```suggestion
            glCode: 'GL Code',
```

File: src/libs/Navigation/linkingConfig/FULL_SCREEN_TO_RHP_MAPPING.ts
Comment (Beamanator) [inline src/libs/Navigation/linkingConfig/FULL_SCREEN_TO_RHP_MAPPING.ts:82]: mayyyyy want to call this `_EDIT_GLCODE`? Maybe?

File: src/libs/Navigation/types.ts
Comment (Beamanator) [inline src/libs/Navigation/types.ts:205]: I'm noticing the other workspace category pages have `backTo` so we might want to allow that here?

Comment (Beamanator) [reply src/libs/Navigation/types.ts:205]: Or maybe that's cuz we're linking to category pages from other places in the app, where we probably won't link to this GL Code edit page from other pages in the app? maybe?

File: src/libs/actions/Policy/Category.ts
Comment (Beamanator) [inline src/libs/actions/Policy/Category.ts:337]: I think we'll want to call this `update*`

File: src/libs/actions/Policy/Category.ts
Comment (Beamanator) [inline src/libs/actions/Policy/Category.ts:351]: IMO we won't need pending stuff, I think this would be offline pattern A, offline no feedback - cuz this is such a tiny field to update, but would be good to confirm in the doc

File: src/pages/workspace/categories/CategorySettingsPage.tsx
Comment (Beamanator) [inline src/pages/workspace/categories/CategorySettingsPage.tsx:141]: Same comment about offline with feedback 🤔  I see that category name has pending fields.... so maybe we would actually want gl code to have it too? probably? maybe need to get confirmation from people in slack, from whoever implemented categories hahaha

**Review by Beamanator — COMMENTED** [review]
obviously 99% looking good, the rest will probably just need to be confirmed in the doc or here :D 

But ya just to confirm, you're planning to add the "relatively controversial changes" from this PR into the doc, right? We'll need approval there before we can merge this :D

File: src/languages/en.ts
Comment (rushatgabhane) [reply src/languages/en.ts:2177]: it's small c in the doc, so i'll stick to that. We follow sentence case

File: src/libs/Navigation/linkingConfig/FULL_SCREEN_TO_RHP_MAPPING.ts
Comment (rushatgabhane) [reply src/libs/Navigation/linkingConfig/FULL_SCREEN_TO_RHP_MAPPING.ts:82]: yeah sure, i'll update this everywhere actually.

File: src/libs/Navigation/types.ts
Comment (rushatgabhane) [reply src/libs/Navigation/types.ts:205]: Yes exactly! I don't see us linking the `gl code` page from somewhere else. In any case, I think we should add this feature only when it's actually being used somewhere.

File: src/libs/actions/Policy/Category.ts
Comment (rushatgabhane) [reply src/libs/actions/Policy/Category.ts:337]: sounds good

File: src/libs/actions/Policy/Category.ts
Comment (rushatgabhane) [reply src/libs/actions/Policy/Category.ts:351]: hmmmm - `Does the user need to know if the action was successful?`

I think yes because the user would go ahead and export CSVs thinking glCode thing works but would get frustrated that it didn't work?? 

image

Comment (Beamanator) [reply src/libs/actions/Policy/Category.ts:351]: I like how there's 2 "Most preferred" options hahaha. Yeah 🤷  IMO it's going to be insanely unlikely that these DON'T get updated successfully and IMO I'd prefer pattern A - cleaner code. But let's see what happens in the doc :D

File: src/libs/Navigation/types.ts
Comment (Beamanator) [reply src/libs/Navigation/types.ts:205]: agreed, ya good plan 👍

File: src/languages/en.ts
Comment (Beamanator) [reply src/languages/en.ts:2177]: minnnnnteresting, then ya i'd go with what's in the HL of the doc 👍

**Comment by rushatgabhane** [comment]
@shubham1206agra please give it one pass while the backend gets deployed

File: src/libs/actions/Policy/Category.ts
Comment (rushatgabhane) [reply src/libs/actions/Policy/Category.ts:376]: this ISSS the old data, right?

File: src/libs/actions/Policy/Category.ts
Comment (Beamanator) [reply src/libs/actions/Policy/Category.ts:348]: Resolving since we're going with pattern B!

File: src/libs/API/parameters/index.ts
Comment (Beamanator) [inline src/libs/API/parameters/index.ts:167]: NAB - damn i wish we could rename all of these to "Policy" for consistency...... But not really appropriate in this PR lol

File: src/libs/actions/Policy/Category.ts
Comment (Beamanator) [inline src/libs/actions/Policy/Category.ts:378]: We should also clear pendingFields, right?

```suggestion
                        pendingAction: null,
                        pendingFields: {
                            glCode: null,
                        },
```

File: src/pages/workspace/categories/CategorySettingsPage.tsx
Comment (Beamanator) [reply src/pages/workspace/categories/CategorySettingsPage.tsx:129]: looks like you already added this, resolving :D

File: src/pages/workspace/categories/CategoryGLCodePage.tsx
Comment (Beamanator) [inline src/pages/workspace/categories/CategoryGLCodePage.tsx:81]: Hmm why not `CONST.FORM_CHARACTER_LIMIT`? I didn't look at the doc recently, but i imagine GL codes are never > 50 characters in length?

**Review by Beamanator — COMMENTED** [review]
Lookin good still!! Only a few tiny comments left :D

File: src/pages/workspace/categories/CategoryGLCodePage.tsx
Comment (rushatgabhane) [reply src/pages/workspace/categories/CategoryGLCodePage.tsx:81]: @Beamanator if im not wrong, in database the limit is 256 chars

File: src/libs/actions/Policy/Category.ts
Comment (Gonals) [inline src/libs/actions/Policy/Category.ts:349]: This should go under "GL Code", so "GL Code": glCode

**Comment by Gonals** [comment]
Backend is merged! We are good to test and review for this one and the payrollcode one

**Comment by rushatgabhane** [comment]
Hold lifted. This can be reviewed and merged!

**Review by shubham1206agra — APPROVED** [review]

**Comment by Gonals** [comment]
Seems like we have some conflicts!

**Comment by Gonals** [comment]
If there is an error, the failed code remains in the field even when I dismiss the error. Let's make sure that gets cleaned up
Screenshot 2024-07-08 at 12 07 52 PM

**Review by Gonals — CHANGES_REQUESTED** [review]
A couple of comments. Looking good otherwise!

File: src/languages/es.ts
Comment (Gonals) [inline src/languages/es.ts:2527]: `Código GL`

`Se produjo un error al actualizar el código GL. Inténtelo nuevamente.`

File: src/pages/workspace/categories/CategoryGLCodePage.tsx
Comment (Beamanator) [reply src/pages/workspace/categories/CategoryGLCodePage.tsx:77]: Merp sorry for not responding here - you def may be right, but the better question to me is: Can GL codes ever get anywhere near 50 characters even? If not, I'd prefer we limit to 50 chars

Comment (Beamanator) [reply src/pages/workspace/categories/CategoryGLCodePage.tsx:77]: @Gonals any opinion here? 🤷

**Review by Beamanator — DISMISSED** [review]
Looks good to me! I had just one comment still about the char validation length for GL codes, but NAB

**Comment by Beamanator** [comment]
Oh @rushatgabhane conflicts 😬

**Comment by rushatgabhane** [comment]
> If there is an error, the failed code remains in the field even when I dismiss the error. Let's make sure that gets cleaned up

Can't repro the bug. I changed the GL code from `abc` to  `123` and the request failed.

You'll notice that the GL code was reverted to `abc`.


https://github.com/Expensify/App/assets/29673073/9a0635a5-32f1-4be2-b8dc-a9a73248126f

**Review by shubham1206agra — APPROVED** [review]

**Comment by Gonals** [comment]
Seems like we have some conflicts!

**Comment by rushatgabhane** [comment]
All resolved @Gonals

**Review by Gonals — APPROVED** [review]

**Comment by lanitochka17** [comment]
This PR is failing because of issue https://github.com/Expensify/App/issues/45429

The issue is reproducible in: All 

https://github.com/user-attachments/assets/bf7bb5fc-c2f0-44ff-993e-1c02238cd5d3

**Pull request merged**
